### PR TITLE
Total review

### PR DESCRIPTION
**Pull request by Tishka17**

### CODE REVIEW

File: src/dishka/dependency_source.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+from .scope import BaseScope
+
+
+class FactoryType(Enum):
+    GENERATOR = "generator"
+    ASYNC_GENERATOR = "async_generator"
+    FACTORY = "factory"
+    ASYNC_FACTORY = "async_factory"
+    VALUE = "value"
```
Comment (pomponchik) [inline src/dishka/dependency_source.py:30]: 🤔

**Review by zhPavel — CHANGES_REQUESTED** [review]

File: src/dishka/container.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    ) -> "ContextWrapper":
+        """
+        Prepare container for entering the inner scope.
+        :param context: Data which will available in inner scope
+        :param with_lock: Whether synchronize dependency cache or not
+        :return: context manager for inner scope
+        """
+        if not self.child_registries:
+            raise ValueError("No child scopes found")
+        return ContextWrapper(self._get_child(context, with_lock))
```
[the remark this answers is not in the document]
Comment (Tishka17) [reply src/dishka/container.py:71]: 1. Lock must be unique per scope. Concurrent scopes do not need to be synchronized. Only concurrent usage of safe scoped-containers. E.g. requesting APP dependency from two concurrent Request scopes should be synchronized
2. I suppose that typical case is to Synchronize APP dependencies but not Request ones.

File: src/dishka/container.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+    def _get_parent(self, dependency_type: Type[T]) -> T:
+        return self.parent_container.get(dependency_type)
```
[the remark this answers is not in the document]
Comment (Tishka17) [reply src/dishka/container.py:74]: I suppose that typical case is to have 2 scopes (though we have more). So let's keep in mind for future

File: src/dishka/async_container.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+            with_lock: bool = False,
+    ):
+        self.registry = registry
+        self.child_registries = child_registries
+        self.context = {type(self): self}
+        if context:
+            self.context.update(context)
+        self.parent_container = parent_container
+        if with_lock:
+            self.lock = Lock()
```
[the remark this answers is not in the document]
Comment (Tishka17) [reply src/dishka/async_container.py]: сделал пока фабрику аргументом, интерфейс вынесу потом

**Pull request closed without merging**